### PR TITLE
Rework TMR and EOI_EXIT_BITMAP updating logic

### PR DESCRIPTION
**Pull request by lyan3**
These patches make TMR set/clear behavior like real hardware, and update EOI_EXIT
bitmap upon changes to related fields of RTEs.

With these changes, we may get rid of current issue to support per_cpu vector in guest
OS. Because only target vlapic TMR bit will be set/cleared.

For EOI_EXIT_BITMAP, we removed the logic dependency from TMR, and update the bitmaps
according to contents of RTEs. And we also optimized the process to go through all
RTEs one-time on a singel cpu for each update.

Tracked-On: #2343

### CODE REVIEW

**Comment by HuiHuangShi** [comment]
2352<br>29 new potential violations to the coding guideline detected. Please help check. Thanks!  <br><br>1.filename:/hypervisor/include/arch/x86/guest/vcpu.h&nbsp;&nbsp;function:vcpu_vlapic&nbsp;&nbsp;offset:302:<br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;reason:Static procedure is not explicitly called in code analysed. : vcpu_vlapic<br><br>2.filename:/hypervisor/arch/x86/guest/vlapic.c&nbsp;&nbsp;function:vlapic_set_tmr&nbsp;&nbsp;offset:469:<br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;reason:<#> (void) missing for discarded return value.<br><br>3.filename:/hypervisor/arch/x86/guest/vlapic.c&nbsp;&nbsp;function:vlapic_set_tmr&nbsp;&nbsp;offset:469:<br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;reason:Pointer not checked for null before use. : tmrptr<br><br>4.filename:/hypervisor/arch/x86/guest/vlapic.c&nbsp;&nbsp;function:vlapic_set_tmr&nbsp;&nbsp;offset:471:<br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;reason:Pointer not checked for null before use. : tmrptr<br><br>5.filename:/hypervisor/arch/x86/guest/vlapic.c&nbsp;&nbsp;function:vlapic_set_tmr&nbsp;&nbsp;offset:471:<br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;reason:<#> (void) missing for discarded return value.<br><br>6.filename:/hypervisor/arch/x86/guest/vlapic.c&nbsp;&nbsp;function:vlapic_reset_tmr&nbsp;&nbsp;offset:485:<br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;reason:No space between if, while, for and expresn.<br><br>7.filename:/hypervisor/arch/x86/guest/vlapic.c&nbsp;&nbsp;function:vlapic_reset_tmr&nbsp;&nbsp;offset:486:<br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;reason:Pointer not checked for null before use. : lapic<br><br>8.filename:/hypervisor/include/arch/x86/guest/vioapic.h&nbsp;&nbsp;function:None&nbsp;&nbsp;offset:102:<br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;reason:No definition in system for prototyped procedure. : vioapic_update_tmr<br><br>9.filename:/hypervisor/dm/vioapic.c&nbsp;&nbsp;function:vioapic_update_eoi_exit&nbsp;&nbsp;offset:179:<br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;reason:Pointer param should be declared pointer to const. : vioapic<br><br>10.filename:/hypervisor/dm/vioapic.c&nbsp;&nbsp;function:vioapic_update_eoi_exit&nbsp;&nbsp;offset:186:<br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;reason:Procedure contains UR data flow anomalies. : vcpu_id<br><br>11.filename:/hypervisor/dm/vioapic.c&nbsp;&nbsp;function:vioapic_update_eoi_exit&nbsp;&nbsp;offset:192:<br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;reason:Inconsistent usage of loop control variable.<br><br>12.filename:/hypervisor/dm/vioapic.c&nbsp;&nbsp;function:vioapic_update_eoi_exit&nbsp;&nbsp;offset:192:<br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;reason:No space after semi colon in for expression.<br><br>13.filename:/hypervisor/dm/vioapic.c&nbsp;&nbsp;function:vioapic_update_eoi_exit&nbsp;&nbsp;offset:192:<br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;reason:For loop initialisation is not simple.<br><br>14.filename:/hypervisor/dm/vioapic.c&nbsp;&nbsp;function:vioapic_update_eoi_exit&nbsp;&nbsp;offset:192:<br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;reason:Expression has side effects. : vcpu_id<br><br>15.filename:/hypervisor/dm/vioapic.c&nbsp;&nbsp;function:vioapic_update_eoi_exit&nbsp;&nbsp;offset:192:<br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;reason:Use of comma operator.<br><br>16.filename:/hypervisor/dm/vioapic.c&nbsp;&nbsp;function:vioapic_update_eoi_exit&nbsp;&nbsp;offset:192:<br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;reason:Pointer not checked for null before use. : vcpu<br><br>17.filename:/hypervisor/dm/vioapic.c&nbsp;&nbsp;function:vioapic_update_eoi_exit&nbsp;&nbsp;offset:192:<br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;reason:Type conversion without cast. :  (enum vm_state and enum vcpu_state): vcpu -> state != VCPU_OFFLINE<br><br>18.filename:/hypervisor/dm/vioapic.c&nbsp;&nbsp;function:vioapic_update_eoi_exit&nbsp;&nbsp;offset:192:<br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;reason:For loop incrementation is not simple.<br><br>19.filename:/hypervisor/dm/vioapic.c&nbsp;&nbsp;function:vioapic_update_eoi_exit&nbsp;&nbsp;offset:192:<br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;reason:No brackets to loop body.<br><br>20.filename:/hypervisor/dm/vioapic.c&nbsp;&nbsp;function:vioapic_update_eoi_exit&nbsp;&nbsp;offset:205:<br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;reason:Logical conjunctions need brackets.<br><br>21.filename:/hypervisor/dm/vioapic.c&nbsp;&nbsp;function:vioapic_update_eoi_exit&nbsp;&nbsp;offset:227:<br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;reason:Expression has side effects. : vcpu_id<br><br>22.filename:/hypervisor/dm/vioapic.c&nbsp;&nbsp;function:vioapic_update_eoi_exit&nbsp;&nbsp;offset:227:<br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;reason:No brackets to loop body.<br><br>23.filename:/hypervisor/dm/vioapic.c&nbsp;&nbsp;function:vioapic_update_eoi_exit&nbsp;&nbsp;offset:227:<br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;reason:Inconsistent usage of loop control variable.<br><br>24.filename:/hypervisor/dm/vioapic.c&nbsp;&nbsp;function:vioapic_update_eoi_exit&nbsp;&nbsp;offset:227:<br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;reason:For loop initialisation is not simple.<br><br>25.filename:/hypervisor/dm/vioapic.c&nbsp;&nbsp;function:vioapic_update_eoi_exit&nbsp;&nbsp;offset:227:<br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;reason:No space after semi colon in for expression.<br><br>26.filename:/hypervisor/dm/vioapic.c&nbsp;&nbsp;function:vioapic_update_eoi_exit&nbsp;&nbsp;offset:227:<br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;reason:For loop incrementation is not simple.<br><br>27.filename:/hypervisor/dm/vioapic.c&nbsp;&nbsp;function:vioapic_update_eoi_exit&nbsp;&nbsp;offset:227:<br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;reason:Use of comma operator.<br><br>28.filename:/hypervisor/dm/vioapic.c&nbsp;&nbsp;function:vioapic_update_eoi_exit&nbsp;&nbsp;offset:227:<br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;reason:Type conversion without cast. :  (enum vm_state and enum vcpu_state): vcpu -> state != VCPU_OFFLINE<br><br>29.filename:/hypervisor/dm/vioapic.c&nbsp;&nbsp;function:vioapic_update_eoi_exit&nbsp;&nbsp;offset:227:<br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;reason:Pointer not checked for null before use. : vcpu<br><br>

**Comment by lyan3** [comment]
- Fixed violation 2~9 and 20;
- violation 1 is false positive;
- the others are cuased by foreach_vcpu macro;

File: hypervisor/arch/x86/guest/vcpu.c [hunk 1]
```diff
@@ -166,6 +166,49 @@ void vcpu_set_guest_msr(struct acrn_vcpu *vcpu, uint32_t msr, uint64_t val)
...
+		exec_vmwrite64(VMX_EOI_EXIT0_FULL, vcpu->arch.eoi_exit_bitmap[0]);
+		exec_vmwrite64(VMX_EOI_EXIT1_FULL, vcpu->arch.eoi_exit_bitmap[1]);
+		exec_vmwrite64(VMX_EOI_EXIT2_FULL, vcpu->arch.eoi_exit_bitmap[2]);
+		exec_vmwrite64(VMX_EOI_EXIT3_FULL, vcpu->arch.eoi_exit_bitmap[3]);
+	}
+	spinlock_release(&(vcpu->arch.lock));
+}
+
+/*
+ * Reset the eoi_exit_bitmaps
```
Comment (JasonChenCJ) [inline hypervisor/arch/x86/guest/vcpu.c]: wrong comments

**Comment by wenlingz** [comment]
@dongyaozu please help approve this PR,thanks

File: hypervisor/arch/x86/guest/vcpu.c (hunk 1 above)
Comment (lyan3) [reply hypervisor/arch/x86/guest/vcpu.c]: Corrected. Thanks.

**Comment by lyan3** [comment]
Hi @wenlingz , 
could you help check the failure? 
I just corrected the comment error Jason pointed out. 
Thanks.

**Comment by HuiHuangShi** [comment]
2352<br>23 new potential violations to the coding guideline detected. Please help check. Thanks!  <br><br>1.filename:/hypervisor/include/arch/x86/guest/vcpu.h&nbsp;&nbsp;function:vcpu_vlapic&nbsp;&nbsp;offset:302:<br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;reason:Static procedure is not explicitly called in code analysed. : vcpu_vlapic<br><br>2.filename:/hypervisor/arch/x86/guest/vlapic.c&nbsp;&nbsp;function:vlapic_set_tmr&nbsp;&nbsp;offset:469:<br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;reason:Pointer not checked for null before use. : tmrptr<br><br>3.filename:/hypervisor/arch/x86/guest/vlapic.c&nbsp;&nbsp;function:vlapic_set_tmr&nbsp;&nbsp;offset:471:<br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;reason:Pointer not checked for null before use. : tmrptr<br><br>4.filename:/hypervisor/arch/x86/guest/vlapic.c&nbsp;&nbsp;function:vlapic_reset_tmr&nbsp;&nbsp;offset:486:<br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;reason:Pointer not checked for null before use. : lapic<br><br>5.filename:/hypervisor/dm/vioapic.c&nbsp;&nbsp;function:vioapic_update_eoi_exit&nbsp;&nbsp;offset:187:<br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;reason:Procedure contains UR data flow anomalies. : vcpu_id<br><br>6.filename:/hypervisor/dm/vioapic.c&nbsp;&nbsp;function:vioapic_update_eoi_exit&nbsp;&nbsp;offset:193:<br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;reason:Inconsistent usage of loop control variable.<br><br>7.filename:/hypervisor/dm/vioapic.c&nbsp;&nbsp;function:vioapic_update_eoi_exit&nbsp;&nbsp;offset:193:<br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;reason:No space after semi colon in for expression.<br><br>8.filename:/hypervisor/dm/vioapic.c&nbsp;&nbsp;function:vioapic_update_eoi_exit&nbsp;&nbsp;offset:193:<br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;reason:For loop initialisation is not simple.<br><br>9.filename:/hypervisor/dm/vioapic.c&nbsp;&nbsp;function:vioapic_update_eoi_exit&nbsp;&nbsp;offset:193:<br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;reason:Expression has side effects. : vcpu_id<br><br>10.filename:/hypervisor/dm/vioapic.c&nbsp;&nbsp;function:vioapic_update_eoi_exit&nbsp;&nbsp;offset:193:<br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;reason:Use of comma operator.<br><br>11.filename:/hypervisor/dm/vioapic.c&nbsp;&nbsp;function:vioapic_update_eoi_exit&nbsp;&nbsp;offset:193:<br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;reason:Pointer not checked for null before use. : vcpu<br><br>12.filename:/hypervisor/dm/vioapic.c&nbsp;&nbsp;function:vioapic_update_eoi_exit&nbsp;&nbsp;offset:193:<br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;reason:Type conversion without cast. :  (enum vm_state and enum vcpu_state): vcpu -> state != VCPU_OFFLINE<br><br>13.filename:/hypervisor/dm/vioapic.c&nbsp;&nbsp;function:vioapic_update_eoi_exit&nbsp;&nbsp;offset:193:<br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;reason:For loop incrementation is not simple.<br><br>14.filename:/hypervisor/dm/vioapic.c&nbsp;&nbsp;function:vioapic_update_eoi_exit&nbsp;&nbsp;offset:193:<br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;reason:No brackets to loop body.<br><br>15.filename:/hypervisor/dm/vioapic.c&nbsp;&nbsp;function:vioapic_update_eoi_exit&nbsp;&nbsp;offset:228:<br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;reason:Expression has side effects. : vcpu_id<br><br>16.filename:/hypervisor/dm/vioapic.c&nbsp;&nbsp;function:vioapic_update_eoi_exit&nbsp;&nbsp;offset:228:<br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;reason:No brackets to loop body.<br><br>17.filename:/hypervisor/dm/vioapic.c&nbsp;&nbsp;function:vioapic_update_eoi_exit&nbsp;&nbsp;offset:228:<br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;reason:Inconsistent usage of loop control variable.<br><br>18.filename:/hypervisor/dm/vioapic.c&nbsp;&nbsp;function:vioapic_update_eoi_exit&nbsp;&nbsp;offset:228:<br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;reason:For loop initialisation is not simple.<br><br>19.filename:/hypervisor/dm/vioapic.c&nbsp;&nbsp;function:vioapic_update_eoi_exit&nbsp;&nbsp;offset:228:<br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;reason:No space after semi colon in for expression.<br><br>20.filename:/hypervisor/dm/vioapic.c&nbsp;&nbsp;function:vioapic_update_eoi_exit&nbsp;&nbsp;offset:228:<br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;reason:For loop incrementation is not simple.<br><br>21.filename:/hypervisor/dm/vioapic.c&nbsp;&nbsp;function:vioapic_update_eoi_exit&nbsp;&nbsp;offset:228:<br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;reason:Use of comma operator.<br><br>22.filename:/hypervisor/dm/vioapic.c&nbsp;&nbsp;function:vioapic_update_eoi_exit&nbsp;&nbsp;offset:228:<br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;reason:Type conversion without cast. :  (enum vm_state and enum vcpu_state): vcpu -> state != VCPU_OFFLINE<br><br>23.filename:/hypervisor/dm/vioapic.c&nbsp;&nbsp;function:vioapic_update_eoi_exit&nbsp;&nbsp;offset:228:<br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;reason:Pointer not checked for null before use. : vcpu<br><br>

**Pull request merged**